numbat/src/typed_ast.rs

```diff
@@ -41,10 +41,22 @@ type DtypeFactorPower = (DTypeFactor, Exponent);
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct DType {
     // Always in canonical form
-    pub factors: Vec<DtypeFactorPower>, // TODO make this private
+    factors: Vec<DtypeFactorPower>,
 }
 
 impl DType {
+    pub fn new(factors: Vec<DtypeFactorPower>) -> Self {
+        Self { factors }
+    }
+
+    pub fn factors(&self) -> &[DtypeFactorPower] {
+        &self.factors
+    }
+
+    pub fn factors_mut(&mut self) -> &mut [DtypeFactorPower] {
+        &mut self.factors
+    }
```

Comment:
I think the idea of the TODO comment was not to get rid of the `pub`, but rather to prevent mutable access to the `factors` member, in order to make sure that invariants of `DType` are always fulfilled (i.e. that `factors` are in canonical form).

Reply:
Gotcha. I just changed a few things around so that a `DType` can be consumed to provide its `factors`, and then can only be created again via `from_factors`, ensuring it's always canonicalized.